### PR TITLE
Add Cross-Origin Resource Sharing (CORS) headers to RestServer

### LINKED ISSUE #654: Enable CORS policy on FSCrawler REST client
I'm trying to create an Angular 7 web app where I can add PDF contents directly into a Elasticsearch service. I created the file uploader in the app and I started a FSCrawler REST client to send the file requests. But when I try it, the web browser pops out this in console:

"Access to XMLHttpRequest at 'http://127.0.0.1:8080/fscrawler' from origin 'http://localhost:4200' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource."

I tried previously to make some opps directly on the Elasticsearch client (creating an index e.g) and I enabled the CORS in my Elasticsearch so it worked. But even with that done, this FSCrawler request doesn't work. 
Anyone can help me please? thx

### PR DESCRIPTION
**Pull request by isaac-ipl**
I thought about making the Access-Control-* headers individual options but decided to just keep it simple and either allow all or not.

Haven't added any new tests - just made sure this passed the existing ones. 

Also the docs will need to be updated to reflect the `enable_cors` setting (default `false`).
```
rest:
  url: "http://127.0.0.1:8080/fscrawler"
  enable_cors: false
```

Closes #654.

### CODE REVIEW

**Comment by dadoonet** [comment]
Hi @isaac-ipl 

Thank you so much for your contribution. Much appreciated.
The build is not passing but this because of something else which I'm going to fix as soon as I can.
I'm going to review your proposal. Stay tuned.

File: cli/src/main/java/fr/pilato/elasticsearch/crawler/fs/cli/FsCrawlerCli.java [hunk 1]
```diff
@@ -27,12 +27,7 @@
 import fr.pilato.elasticsearch.crawler.fs.framework.FsCrawlerUtil;
 import fr.pilato.elasticsearch.crawler.fs.framework.MetaFileHandler;
 import fr.pilato.elasticsearch.crawler.fs.rest.RestServer;
-import fr.pilato.elasticsearch.crawler.fs.settings.Elasticsearch;
-import fr.pilato.elasticsearch.crawler.fs.settings.Fs;
-import fr.pilato.elasticsearch.crawler.fs.settings.FsCrawlerValidator;
-import fr.pilato.elasticsearch.crawler.fs.settings.FsSettings;
-import fr.pilato.elasticsearch.crawler.fs.settings.FsSettingsFileHandler;
-import fr.pilato.elasticsearch.crawler.fs.settings.FsSettingsParser;
+import fr.pilato.elasticsearch.crawler.fs.settings.*;
```
Comment (dadoonet) [inline cli/src/main/java/fr/pilato/elasticsearch/crawler/fs/cli/FsCrawlerCli.java]: Could you keep the explicit imports?

File: integration-tests/it-common/src/main/java/fr/pilato/elasticsearch/crawler/fs/test/integration/AbstractFsCrawlerITCase.java [hunk 2]
```diff
@@ -23,11 +23,7 @@
 import fr.pilato.elasticsearch.crawler.fs.beans.FsJobFileHandler;
 import fr.pilato.elasticsearch.crawler.fs.client.ESSearchRequest;
 import fr.pilato.elasticsearch.crawler.fs.framework.TimeValue;
-import fr.pilato.elasticsearch.crawler.fs.settings.Elasticsearch;
-import fr.pilato.elasticsearch.crawler.fs.settings.Fs;
-import fr.pilato.elasticsearch.crawler.fs.settings.FsSettings;
-import fr.pilato.elasticsearch.crawler.fs.settings.Server;
-import fr.pilato.elasticsearch.crawler.fs.settings.ServerUrl;
+import fr.pilato.elasticsearch.crawler.fs.settings.*;
```
Comment (dadoonet) [inline integration-tests/it-common/src/main/java/fr/pilato/elasticsearch/crawler/fs/test/integration/AbstractFsCrawlerITCase.java]: Could you keep the explicit imports?

File: cli/src/main/java/fr/pilato/elasticsearch/crawler/fs/cli/FsCrawlerCli.java [hunk 3]
```diff
@@ -229,6 +224,7 @@ public static void main(String[] args) throws Exception {
                 fsSettings = FsSettings.builder(commands.jobName.get(0))
                         .setFs(Fs.DEFAULT)
                         .setElasticsearch(Elasticsearch.DEFAULT())
+                        .setRest(Rest.DEFAULT)
```
Comment (dadoonet) [inline cli/src/main/java/fr/pilato/elasticsearch/crawler/fs/cli/FsCrawlerCli.java]: I don't think this is needed. We are setting that elsewhere (in `FsCrawlerValidator`) if FsCrawler is started with the rest mode.

File: integration-tests/it-common/src/main/java/fr/pilato/elasticsearch/crawler/fs/test/integration/AbstractITCase.java [hunk 4]
```diff
@@ -117,6 +118,8 @@
     final static int testRestPort = Integer.parseInt(System.getProperty("tests.rest.port", DEFAULT_TEST_REST_PORT.toString()));
     static Elasticsearch elasticsearchWithSecurity;
 
+    static Rest restConfig;
```
Comment (dadoonet) [inline integration-tests/it-common/src/main/java/fr/pilato/elasticsearch/crawler/fs/test/integration/AbstractITCase.java]: I don't think we need this. It's never used.

File: settings/src/main/java/fr/pilato/elasticsearch/crawler/fs/settings/Rest.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
+package fr.pilato.elasticsearch.crawler.fs.settings;
+
+import java.util.Objects;
+
+public class Rest {
```
Comment (dadoonet) [inline settings/src/main/java/fr/pilato/elasticsearch/crawler/fs/settings/Rest.java]: This change makes sense to me.
Could you change the javadoc in `ServerUrl` class?
I wrote:

```
/**
 * This class represents a ServerUrl which is basically just a String which
 * can be either a url or a cloud id.
 * This is used in Rest and Elasticsearch.Node classes.
 */
```

The last line is not true anymore and only applies to `Elasticsearch.Node` class.

File: settings/src/test/java/fr/pilato/elasticsearch/crawler/fs/settings/FsSettingsParserTest.java [hunk 6]
```diff
@@ -77,7 +77,11 @@
             .setProtocol("SSH")
             .setPemPath("/path/to/pemfile")
             .build();
-    private static final ServerUrl REST_FULL = new ServerUrl("http://127.0.0.1:8080/fscrawler");
+    private static final Rest REST_EMPTY = Rest.builder().build();
```
Comment (dadoonet) [inline settings/src/test/java/fr/pilato/elasticsearch/crawler/fs/settings/FsSettingsParserTest.java]: We are not using `REST_EMPTY`. Could you remove that or add a test that uses it?

**Comment by isaac-ipl** [comment]
Hi @dadoonet thanks for the feedback, I have made the requested changes. 

I was planning to implement some tests that would compare the RestServer response headers against expected, but unfortunately won't be able to commit to doing that at this point in time.

File: rest/src/main/java/fr/pilato/elasticsearch/crawler/fs/rest/CORSFilter.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+@Provider
+public class CORSFilter implements ContainerResponseFilter {
+    private Rest rest;
+
+    public CORSFilter(Rest rest) {
+        this.rest = rest;
+    }
+
+    @Override
+    public void filter(ContainerRequestContext request, ContainerResponseContext response) throws IOException {
```
[the remark this answers is not in the document]
Comment (isaac-ipl) [reply rest/src/main/java/fr/pilato/elasticsearch/crawler/fs/rest/CORSFilter.java]: You're right, it's not applicable. It was autocompleted when I created the class.

File: settings/src/main/java/fr/pilato/elasticsearch/crawler/fs/settings/FsCrawlerValidator.java [hunk 8]
```diff
@@ -108,8 +108,8 @@ public static boolean validateSettings(Logger logger, FsSettings settings, boole
 
         // Check that REST settings are available when we start with rest option
         if (rest && settings.getRest() == null) {
-            logger.warn("`rest` settings are not defined. Falling back to default: [{}].", FsSettings.REST_DEFAULT);
-            settings.setRest(FsSettings.REST_DEFAULT);
+            logger.warn("`rest` settings are not defined. Falling back to default: [{}].", Rest.URL_DEFAULT);
```
[the remark this answers is not in the document]
Comment (isaac-ipl) [reply settings/src/main/java/fr/pilato/elasticsearch/crawler/fs/settings/FsCrawlerValidator.java]: Hmm, I don't have any issues against or for it. But I was basing it off of the other settings classes, like how Fs has ``DEFAULT_DIR`` and ``DEFAULT`` public and referenced. Up to you :)

File: settings/src/test/java/fr/pilato/elasticsearch/crawler/fs/settings/FsSettingsParserTest.java [hunk 9]
```diff
@@ -77,7 +77,10 @@
             .setProtocol("SSH")
             .setPemPath("/path/to/pemfile")
             .build();
-    private static final ServerUrl REST_FULL = new ServerUrl("http://127.0.0.1:8080/fscrawler");
+    private static final Rest REST_FULL = Rest.builder()
```
[the remark this answers is not in the document]
Comment (isaac-ipl) [reply settings/src/test/java/fr/pilato/elasticsearch/crawler/fs/settings/FsSettingsParserTest.java]: True that it's the same thing now, but in future there may be configurables that aren't always defined by default. Would keeping it as is make it easier to understand and modify?

**Comment by isaac-ipl** [comment]
Thanks, I'm no expert either but from my understanding it's needed only to let browsers know they are allowed to access a resource from a different origin. Otherwise the browser will prevent the request. It's commonly needed with front-ends that make AJAX requests to resources of different origins.

Here's a couple of good resources: 
https://stackoverflow.com/a/10636765/1235586
https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#Simple_requests

and a useful graphic: https://upload.wikimedia.org/wikipedia/commons/c/ca/Flowchart_showing_Simple_and_Preflight_XHR.svg

As for testing, I simply checked the REST response headers:
```
# CORS disabled
$ curl -I http://127.0.0.1:8080/fscrawler
HTTP/1.1 200 OK
Content-Length: 735
Content-Type: application/json
```

```
# CORS enabled
$ curl -I http://127.0.0.1:8080/fscrawler
HTTP/1.1 200 OK
Access-Control-Allow-Origin: *
Access-Control-Allow-Headers: origin, content-type, accept, authorization
Access-Control-Allow-Credentials: true
Access-Control-Allow-Methods: GET, PATCH, POST, PUT, DELETE, OPTIONS, HEAD
Content-Length: 734
Content-Type: application/json
```

You can also test it interactively with a fairly simple front-end like: https://www.test-cors.org/ (I find it easier to just look at the browser developer console and check for errors due to CORS).

I have updated the docs you pointed out.

File: settings/src/main/java/fr/pilato/elasticsearch/crawler/fs/settings/FsCrawlerValidator.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (dadoonet) [reply settings/src/main/java/fr/pilato/elasticsearch/crawler/fs/settings/FsCrawlerValidator.java]: Ok. Let's keep it your way then.

File: settings/src/test/java/fr/pilato/elasticsearch/crawler/fs/settings/FsSettingsParserTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (dadoonet) [reply settings/src/test/java/fr/pilato/elasticsearch/crawler/fs/settings/FsSettingsParserTest.java]: Ok. Works for me.

**Comment by dadoonet** [comment]
Awesome. Thank you so much for this!

**Pull request merged**